sky/serve/load_balancer.py

```diff
@@ -77,23 +78,85 @@ def _sync_with_controller(self):
                         ready_replica_urls)
             time.sleep(constants.LB_CONTROLLER_SYNC_INTERVAL_SECONDS)
 
-    async def _redirect_handler(self, request: fastapi.Request):
-        self._request_aggregator.add(request)
-        ready_replica_url = self._load_balancing_policy.select_replica(request)
-
-        if ready_replica_url is None:
-            raise fastapi.HTTPException(status_code=503,
-                                        detail='No ready replicas. '
-                                        'Use "sky serve status [SERVICE_NAME]" '
-                                        'to check the replica status.')
+    async def _proxy_request_to(
+            self, url: str,
+            request: fastapi.Request) -> Optional[fastapi.responses.Response]:
+        """Proxy the request to the specified URL.
 
-        path = f'http://{ready_replica_url}{request.url.path}'
-        logger.info(f'Redirecting request to {path}')
-        return fastapi.responses.RedirectResponse(url=path)
+        Returns:
+            The response from the endpoint replica. None if anything goes wrong.
+        """
+        method = request.method
+        headers = {key: value for key, value in request.headers.items()}
+        body = await request.body()
+        path = f'http://{url}{request.url.path}'
+        logger.info(f'Proxy request to {path}')
+        try:
+
+            async def stream_response():
+                async with httpx.AsyncClient() as client:
+                    async with client.stream(method,
+                                             path,
+                                             headers=headers,
+                                             content=body) as response:
+                        response.raise_for_status()
+                        # TODO(tian): Hacky. Investigate a way to not directly
+                        # yielding the response status code and headers.
+                        yield response.status_code
+                        yield dict(response.headers)
```

Comment:
Could we elaborate on why we are yielding this two fields?

Reply:
Since we need to construct the httpx client inside the `stream_response()` function, otherwise it will cause the connection closed bug i mentioned before. While at the same time we need the return status code and headers when constructing the `fastapi.responses.StreamingResponse`, so we have to yield it first. Do you have ideas on how to make this prettier?

Reply:
Should we check out this: https://github.com/tiangolo/fastapi/issues/1788#issuecomment-1071222163?

Reply:
This seems to work well! Lemme run the smoke tests to double check 🙌